infoschema/cluster.go

```diff
@@ -0,0 +1,104 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infoschema
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/domain/infosync"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util"
+)
+
+// Cluster table list, attention:
+// 1. the table name should be upper case.
+// 2. clusterTableName should equal to "TIDB_CLUSTER_" + memTableTableName.
+const (
+	clusterTableSlowLog     = "TIDB_CLUSTER_SLOW_QUERY"
+	clusterTableProcesslist = "TIDB_CLUSTER_PROCESSLIST"
+)
+
+// memTableToClusterTableMap means add memory table to cluster table.
+var memTableToClusterTableMap = map[string]string{
+	tableSlowLog:     clusterTableSlowLog,
+	tableProcesslist: clusterTableProcesslist,
+}
+
+// clusterTableMap is the cluster table map.
+// It will be initialized in `init` function.
+var clusterTableMap = map[string]struct{}{}
+
+func init() {
+	var addrCol = columnInfo{"ADDRESS", mysql.TypeVarchar, 64, 0, nil, nil}
+	for memTableName, clusterMemTableName := range memTableToClusterTableMap {
+		memTableCols := tableNameToColumns[memTableName]
+		if len(memTableCols) == 0 {
+			continue
+		}
+		cols := make([]columnInfo, 0, len(memTableCols)+1)
+		cols = append(cols, memTableCols...)
+		cols = append(cols, addrCol)
+		tableNameToColumns[clusterMemTableName] = cols
+		clusterTableMap[clusterMemTableName] = struct{}{}
+	}
+}
+
+// isClusterTableByName used to check whether the table is a cluster memory table.
+func isClusterTableByName(dbName, tableName string) bool {
+	dbName = strings.ToLower(dbName)
+	switch dbName {
+	case util.InformationSchemaLowerName, util.PerformanceSchemaLowerName:
+		break
+	default:
+		return false
+	}
+	tableName = strings.ToUpper(tableName)
+	_, ok := clusterTableMap[tableName]
+	return ok
+}
+
+// IsClusterTable used to check whether the table is a cluster memory table.
+func IsClusterTable(tp table.Type) bool {
+	return tp == table.ClusterTable
+}
+
+func getClusterMemTableRows(ctx sessionctx.Context, tableName string) (rows [][]types.Datum, err error) {
+	tableName = strings.ToUpper(tableName)
+	switch tableName {
+	case clusterTableSlowLog:
+		rows, err = dataForSlowLog(ctx)
+	case clusterTableProcesslist:
+		rows = dataForProcesslist(ctx)
+	default:
+		err = errors.Errorf("unknown cluster table: %v", tableName)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return appendClusterColumnsToRows(rows), nil
+}
+
+func appendClusterColumnsToRows(rows [][]types.Datum) [][]types.Datum {
+	serverInfo := infosync.GetServerInfo()
+	addr := serverInfo.IP + ":" + strconv.FormatUint(uint64(serverInfo.StatusPort), 10)
```

Comment:
fmt.Sprinf("%s:%s", x, x) ?

Reply:
em.... the performance of `fmt.Sprinf` is bad, so i don't like it...

Reply:
Why use `status port`?